crypto/evp/evp_lib.c

```diff
@@ -1227,3 +1229,299 @@ EVP_PKEY *EVP_PKEY_Q_keygen(OSSL_LIB_CTX *libctx, const char *propq,
+     * smdalgs->lock held on success. */
+    digests = sk_EVP_MD_new(md_comp);
+    if (digests == NULL)
+        goto err;
+
+    callback_info.libctx = libctx;
+    callback_info.digests = digests;
+
+    if (!CONF_parse_list(value, ':', 1,
+                         (int (*)(const char *, int, void *)) add_signature_md,
+                         &callback_info))
+        goto err;
+
+    /* Pre-sort the digests, since we want to be able to run sk_EVP_MD_find()
+     * under a read lock, but sk_EVP_MD_find() will modify unsorted stacks. */
+    sk_EVP_MD_sort(digests);
+
+    /* Lock and swap the list of accepted digests; delay freeing the old list
+     * of digests until after the lock has been released to keep the critical
+     * section short. */
+    if (!CRYPTO_THREAD_write_lock(smdalgs->lock))
+        goto err;
+
+    switch (usecase) {
+    case EVP_SIGNATURE_MD_ALGORITHMS_SIGNING:
+        digests_p = &smdalgs->signing;
+        break;
+    case EVP_SIGNATURE_MD_ALGORITHMS_VERIFICATION:
+        digests_p = &smdalgs->verification;
+        break;
+    default:
+        ERR_raise(ERR_LIB_EVP, ERR_R_PASSED_INVALID_ARGUMENT);
+        goto unlock_err;
+    }
+
+    old_digests = *digests_p;
+    *digests_p = digests;
+    digests = NULL;
+    CRYPTO_THREAD_unlock(smdalgs->lock);
+
+    sk_EVP_MD_pop_free(old_digests, EVP_MD_free);
+
+    return 1;
+
+unlock_err:
+    CRYPTO_THREAD_unlock(smdalgs->lock);
+
+err:
+    sk_EVP_MD_pop_free(digests, EVP_MD_free);
+    return 0;
+}
+
+int EVP_signature_md_algorithm_set(OSSL_LIB_CTX *libctx, int usecase,
+                                   EVP_MD *md, int allow)
+{
+    STACK_OF(EVP_MD) *digests = NULL;
+    int ret = 0;
+
+    OSSL_SIGNATURE_MD_ALGORITHMS *smdalgs = ossl_lib_ctx_get_data(
+        libctx, OSSL_LIB_CTX_SIGNATURE_MD_ALGORITHMS_INDEX);
+    if (smdalgs == NULL)
+        return 0;
+    if (md == NULL) {
+        ERR_raise(ERR_LIB_EVP, ERR_R_PASSED_INVALID_ARGUMENT);
+        return 0;
+    }
+
+    if (!CRYPTO_THREAD_write_lock(smdalgs->lock))
+        return 0;
+
+    switch (usecase) {
+    case EVP_SIGNATURE_MD_ALGORITHMS_SIGNING:
+        digests = smdalgs->signing;
+        break;
+    case EVP_SIGNATURE_MD_ALGORITHMS_VERIFICATION:
+        digests = smdalgs->verification;
+        break;
+    default:
+        ERR_raise(ERR_LIB_EVP, ERR_R_PASSED_INVALID_ARGUMENT);
+        goto unlock;
+    }
+
+    ret = signature_md_algorithm_set(digests, md, allow);
+
+    /* Pre-sort the digests, since we want to be able to run sk_EVP_MD_find()
+     * under a read lock, but sk_EVP_MD_find() will modify unsorted stacks. */
+    sk_EVP_MD_sort(digests);
+
+unlock:
+    CRYPTO_THREAD_unlock(smdalgs->lock);
+
+    return ret;
+}
+
+#endif /* !defined(FIPS_MODULE) */
+
+int EVP_signature_md_algorithm_allowed(OSSL_LIB_CTX *libctx, int usecase, const EVP_MD *md)
+{
+    STACK_OF(EVP_MD) *digests = NULL;
+    int ret = -1;
+
+    OSSL_SIGNATURE_MD_ALGORITHMS *smdalgs = ossl_lib_ctx_get_data(
```

Comment:
Yes, I refactored libctx in #17881 to not use CRYPTO_EX_DATA. `ossl_lib_ctx_get_data()` is now implemented using a switch statement. This PR adds a new item to `ossl_lib_ctx_get_data` which looks fine to me.

If you mean `smdalgs->lock`, looking it at briefly I would assume it would be necessary. In the OSSL_LIB_CTX case the issue was caused by unnecessary use of CRYPTO_EX_DATA.